File: Raven.Client.Java/src/main/java/net/ravendb/client/linq/IRavenQueryable.java
package net.ravendb.client.linq;

import net.ravendb.abstractions.basic.Reference;
import net.ravendb.abstractions.json.linq.RavenJToken;
import net.ravendb.client.LinqExtensionsQueryable;
import net.ravendb.client.RavenQueryStatistics;
import net.ravendb.client.document.DocumentQueryCustomizationFactory;
import net.ravendb.client.indexes.AbstractTransformerCreationTask;
import net.ravendb.client.spatial.SpatialCriteria;

import com.mysema.query.types.Path;

/**
 * An implementation of IOrderedQueryable with Raven specific operation
 *
 * @param <T>
 */
public interface IRavenQueryable<T> extends IOrderedQueryable<T>, LinqExtensionsQueryable<T> {
  /**
   * Provide statistics about the query, such as total count of matching records
   * @param stats
   * @return
   */
  IRavenQueryable<T> statistics(Reference<RavenQueryStatistics> stats);

  /**
   * Customizes the query using the specified action
   * @param action
   * @return
   */
  IRavenQueryable<T> customize(DocumentQueryCustomizationFactory customizationFactory);

  /**
   * Specifies a result transformer to use on the results
   * @param transformerClazz
   * @param resultClass
   * @return
   */
  <S> IRavenQueryable<S> transformWith(String transformerName, Class<S> resultClass);


  /**
   * Specifies a result transformer to use on the results
   * @param transformerClazz
   * @param resultClass
   * @return
   */
  <S> IRavenQueryable<S> transformWith(Class<? extends AbstractTransformerCreationTask> transformerClazz, Class<S> resultClass);

  /**
   * Inputs a key and value to the query (accessible by the transformer)
   * @param name
   * @param value
   * @return
   */
  IRavenQueryable<T> addQueryInput(String name, RavenJToken value);

  /**
   * Inputs a key and value to the query (accessible by the transformer)
   * @param name
   * @param value
   * @return
   */
  IRavenQueryable<T> addQueryInput(String name, Object value);

  IRavenQueryable<T> spatial(Path<?> path, SpatialCriteria criteria);

  /**
   * Returns distinct results
   * @return
   */
  IRavenQueryable<T> distinct();

}

File: src/Raven.NewClient/Linq/IRavenQueryable.cs
//-----------------------------------------------------------------------
// <copyright file="IRavenQueryable.cs" company="Hibernating Rhinos LTD">
//     Copyright (c) Hibernating Rhinos LTD. All rights reserved.
// </copyright>
//-----------------------------------------------------------------------
using System;
using System.Linq;
using System.Linq.Expressions;
using Raven.Abstractions.Indexing;
using Raven.NewClient.Client.Indexes;
using Raven.NewClient.Client.Spatial;
using Raven.NewClient.Json.Linq;

namespace Raven.NewClient.Client.Linq
{
    /// <summary>
    /// An implementation of <see cref="IOrderedQueryable{T}"/> with Raven specific operation
    /// </summary>
    public interface IRavenQueryable<T> : IOrderedQueryable<T>
    {
        /// <summary>
        /// Provide statistics about the query, such as duration, total number of results, staleness information, etc.
        /// </summary>
        IRavenQueryable<T> Statistics(out RavenQueryStatistics stats);

        /// <summary>
        /// Customizes the query using the specified action
        /// </summary>
        IRavenQueryable<T> Customize(Action<IDocumentQueryCustomization> action);

        /// <summary>
        /// Specifies a result transformer to use on the results
        /// </summary>
        IRavenQueryable<TResult> TransformWith<TTransformer, TResult>() where TTransformer : AbstractTransformerCreationTask, new();

        /// <summary>
        /// Specifies a result transformer name to use on the results
        /// </summary>
        IRavenQueryable<TResult> TransformWith<TResult>(string transformerName);

        /// <summary>
        /// Inputs a key and value to the query (accessible by the transformer)
        /// </summary>
        [Obsolete("Use AddTransformerParameter instead.")]
        IRavenQueryable<T> AddQueryInput(string name, RavenJToken value);

        /// <summary>
        /// Inputs a key and value to the query (accessible by the transformer)
        /// </summary>
        IRavenQueryable<T> AddTransformerParameter(string name, RavenJToken value);

        IRavenQueryable<T> Spatial(Expression<Func<T, object>> path, Func<SpatialCriteriaFactory, SpatialCriteria> clause);

        IRavenQueryable<T> OrderByDistance(SpatialSort sortParamsClause);

        /// <summary>
        /// Holds the original query type only when TransformWith is invoked otherwise null.
        /// </summary>
        Type OriginalQueryType { get; set; }
    }
}
